File: apps/cli/src/commands/verify.rs
use crate::cli::{Cli, VerifyArgs};
use crate::commands::{deploy, detect, interact, target};
use crate::error::{AppError, AppResult};
use crate::output::{self, Meta};
use serde::Serialize;
use std::process::Command;

#[derive(Debug, Serialize)]
struct VerifyData {
    target: String,
    contract: String,
    contract_id: String,
    project_root: String,
    address: String,
    chain: Option<String>,
    verifier: Option<String>,
    show_standard_json_input: bool,
    status: String,
    stdout: String,
    stderr: String,
}

pub fn run(cli: &Cli, args: &VerifyArgs) -> AppResult<()> {
    let data = verify_data(cli, args)?;
    if cli.json {
        let mut meta = Meta::new("verify");
        meta.project_root = Some(data.project_root.clone());
        output::print_json(data, meta)
    } else if data.status == "success" {
        if !data.stdout.trim().is_empty() {
            print!("{}", data.stdout);
        }
        if !data.stderr.trim().is_empty() {
            eprint!("{}", data.stderr);
        }
        Ok(())
    } else {
        Err(AppError::user(
            "verify_failed",
            "forge verify-contract failed.",
            Some(if data.stderr.trim().is_empty() {
                data.stdout
            } else {
                data.stderr
            }),
        ))
    }
}

fn verify_data(cli: &Cli, args: &VerifyArgs) -> AppResult<VerifyData> {
    if args.constructor_args.is_some() && args.constructor_args_path.is_some() {
        return Err(AppError::user(
            "verify_constructor_args_conflict",
            "`verify` accepts only one of `--constructor-args` or `--constructor-args-path`.",
            Some(
                "Pass raw constructor args directly, or pass a file path, but not both."
                    .to_string(),
            ),
        ));
    }

    let resolved = target::resolve(cli, Some(&args.target))?;
    deploy::run_forge_build(&resolved.project_root)?;
    let contract_id = deploy::contract_identifier(&resolved)?;
    let network = detect::active_network(cli)?;
    let address = verify_address(cli, args)?;
    let chain = args
        .chain
        .clone()
        .or_else(|| network.chain_id.map(|chain_id| chain_id.to_string()));

    let mut command = Command::new("forge");
    command
        .arg("verify-contract")
        .arg(&address)
        .arg(&contract_id)
        .arg("--root")
        .arg(&resolved.project_root)
        .arg("--rpc-url")
        .arg(&network.rpc_url)
        .arg("--color")
        .arg("never");

    if let Some(chain) = &chain {
        command.arg("--chain").arg(chain);
    }
    if let Some(verifier) = &args.verifier {
        command.arg("--verifier").arg(verifier);
    }
    if let Some(verifier_url) = &args.verifier_url {
        command.arg("--verifier-url").arg(verifier_url);
    }
    if let Some(verifier_api_key) = &args.verifier_api_key {
        command.arg("--verifier-api-key").arg(verifier_api_key);
    }
    if let Some(etherscan_api_key) = &args.etherscan_api_key {
        command.arg("--etherscan-api-key").arg(etherscan_api_key);
    }
    if let Some(constructor_args) = &args.constructor_args {
        command.arg("--constructor-args").arg(constructor_args);
    }
    if let Some(constructor_args_path) = &args.constructor_args_path {
        command
            .arg("--constructor-args-path")
            .arg(constructor_args_path);
    }
    if args.guess_constructor_args {
        command.arg("--guess-constructor-args");
    }
    if args.watch {
        command.arg("--watch");
    }
    if args.show_standard_json_input {
        command.arg("--show-standard-json-input");
    }

    let output = command.output().map_err(|err| {
        AppError::user(
            "verify_failed",
            format!("Failed to run forge verify-contract: {err}"),
            Some("Install Foundry and make sure `forge` is on PATH.".to_string()),
        )
    })?;

    Ok(VerifyData {
        target: args.target.clone(),
        contract: resolved.contract_name,
        contract_id,
        project_root: resolved.project_root.display().to_string(),
        address,
        chain,
        verifier: args.verifier.clone(),
        show_standard_json_input: args.show_standard_json_input,
        status: if output.status.success() {
            "success".to_string()
        } else {
            "failed".to_string()
        },
        stdout: String::from_utf8_lossy(&output.stdout).to_string(),
        stderr: String::from_utf8_lossy(&output.stderr).to_string(),
    })
}

fn verify_address(cli: &Cli, args: &VerifyArgs) -> AppResult<String> {
    if let Some(address) = &args.address {
        return Ok(address.clone());
    }

    interact::context(cli, &args.target)
        .map(|context| context.address)
        .map_err(|err| {
            AppError::user(
                "verify_address_required",
                "No verification address was provided.",
                Some(format!(
                    "Pass `--address <address>` or deploy the target first. Last lookup error: {}",
                    err.message()
                )),
            )
        })
}

File: packages/cli/src/commands/verify.ts
import { ProjectError, resolveTarget } from "@consol/core";
import { runForgeBuild, runForgeVerifyContract } from "@consol/foundry";
import { createSuccessEnvelope } from "@consol/protocol";
import type { GlobalArgs } from "../args";
import type { CliEnv, CliResult } from "../main";
import { VERSION } from "../version";
import { contractIdentifier } from "./contract-id";
import { resolveCliReadNetworkRuntime } from "./network-runtime";

export type VerifyData = {
  readonly target: string;
  readonly contract: string;
  readonly contract_id: string;
  readonly project_root: string;
  readonly address: string;
  readonly chain: string | null;
  readonly verifier: string | null;
  readonly show_standard_json_input: boolean;
  readonly status: "success" | "failed";
  readonly stdout: string;
  readonly stderr: string;
};

export type RunVerifyCommandInput = {
  readonly globals: GlobalArgs;
  readonly commandArgs: readonly string[];
  readonly cwd: string;
  readonly env: CliEnv;
};

type VerifyOptions = {
  readonly target: string;
  readonly address?: string;
  readonly chain?: string;
  readonly verifier?: string;
  readonly verifierUrl?: string;
  readonly verifierApiKey?: string;
  readonly etherscanApiKey?: string;
  readonly constructorArgs?: string;
  readonly constructorArgsPath?: string;
  readonly guessConstructorArgs: boolean;
  readonly watch: boolean;
  readonly showStandardJsonInput: boolean;
};

export async function runVerifyCommand(input: RunVerifyCommandInput): Promise<CliResult> {
  const options = parseVerifyOptions(input.commandArgs);
  if (options.constructorArgs !== undefined && options.constructorArgsPath !== undefined) {
    throw new ProjectError({
      code: "verify_constructor_args_conflict",
      message: "`verify` accepts only one of `--constructor-args` or `--constructor-args-path`.",
      hint: "Pass raw constructor args directly, or pass a file path, but not both.",
    });
  }

  const resolved = resolveTarget({
    cwd: input.cwd,
    target: options.target,
    ...(input.globals.project === undefined ? {} : { projectRoot: input.globals.project }),
  });
  const build = await runForgeBuild({
    cwd: resolved.projectRoot,
    projectRoot: resolved.projectRoot,
    env: input.env,
  });
  if (!build.ok) {
    return { exitCode: 1, stdout: "", stderr: "Foundry build failed before verify.\n" };
  }

  const network = await resolveCliReadNetworkRuntime({ globals: input.globals, cwd: resolved.projectRoot, env: input.env });
  const address = options.address ?? missingAddress();
  const chain = options.chain ?? (network.meta.chain_id === null ? undefined : String(network.meta.chain_id));
  const contractId = contractIdentifier(resolved);
  const verify = await runForgeVerifyContract({
    cwd: resolved.projectRoot,
    projectRoot: resolved.projectRoot,
    env: input.env,
    address,
    contractId,
    rpcUrl: network.rpc_url,
    ...(chain === undefined ? {} : { chain }),
    ...(options.verifier === undefined ? {} : { verifier: options.verifier }),
    ...(options.verifierUrl === undefined ? {} : { verifierUrl: options.verifierUrl }),
    ...(options.verifierApiKey === undefined ? {} : { verifierApiKey: options.verifierApiKey }),
    ...(options.etherscanApiKey === undefined ? {} : { etherscanApiKey: options.etherscanApiKey }),
    ...(options.constructorArgs === undefined ? {} : { constructorArgs: options.constructorArgs }),
    ...(options.constructorArgsPath === undefined ? {} : { constructorArgsPath: options.constructorArgsPath }),
    guessConstructorArgs: options.guessConstructorArgs,
    watch: options.watch,
    showStandardJsonInput: options.showStandardJsonInput,
  });

  const data: VerifyData = {
    target: options.target,
    contract: resolved.contractName,
    contract_id: contractId,
    project_root: resolved.projectRoot,
    address,
    chain: chain ?? null,
    verifier: options.verifier ?? null,
    show_standard_json_input: options.showStandardJsonInput,
    status: verify.ok ? "success" : "failed",
    stdout: verify.stdout,
    stderr: verify.stderr,
  };

  if (input.globals.json || input.commandArgs.includes("--json")) {
    const envelope = createSuccessEnvelope({
      data,
      meta: {
        version: VERSION,
        command: "verify",
        project_root: data.project_root,
      },
    });
    return { exitCode: 0, stdout: `${JSON.stringify(envelope, null, 2)}\n`, stderr: "" };
  }

  return verify.ok
    ? { exitCode: 0, stdout: data.stdout, stderr: data.stderr }
    : { exitCode: 1, stdout: "", stderr: "forge verify-contract failed.\n" };
}

function parseVerifyOptions(commandArgs: readonly string[]): VerifyOptions {
  let target: string | undefined;
  let address: string | undefined;
  let chain: string | undefined;
  let verifier: string | undefined;
  let verifierUrl: string | undefined;
  let verifierApiKey: string | undefined;
  let etherscanApiKey: string | undefined;
  let constructorArgs: string | undefined;
  let constructorArgsPath: string | undefined;
  let guessConstructorArgs = false;
  let watch = false;
  let showStandardJsonInput = false;

  for (let index = 0; index < commandArgs.length; index += 1) {
    const arg = commandArgs[index];
    if (arg === undefined || arg === "--json") {
      continue;
    }

    if (!arg.startsWith("--") && target === undefined) {
      target = arg;
      continue;
    }

    if (arg === "--guess-constructor-args") {
      guessConstructorArgs = true;
      continue;
    }
    if (arg === "--watch") {
      watch = true;
      continue;
    }
    if (arg === "--show-standard-json-input") {
      showStandardJsonInput = true;
      continue;
    }

    const value = commandArgs[index + 1];
    if (value === undefined) {
      throw new ProjectError({
        code: "missing_flag_value",
        message: `Missing value for ${arg}.`,
        hint: `Pass a value after ${arg}.`,
      });
    }

    switch (arg) {
      case "--address":
        address = value;
        break;
      case "--chain":
        chain = value;
        break;
      case "--verifier":
        verifier = value;
        break;
      case "--verifier-url":
        verifierUrl = value;
        break;
      case "--verifier-api-key":
        verifierApiKey = value;
        break;
      case "--etherscan-api-key":
        etherscanApiKey = value;
        break;
      case "--constructor-args":
        constructorArgs = value;
        break;
      case "--constructor-args-path":
        constructorArgsPath = value;
        break;
      default:
        throw new ProjectError({
          code: "verify_arg_unsupported",
          message: `Unsupported verify argument: ${arg}`,
          hint: "Use `consol verify <target> --address <address>`.",
        });
    }
    index += 1;
  }

  if (target === undefined) {
    throw new ProjectError({
      code: "verify_target_required",
      message: "Missing target for verify.",
      hint: "Use `consol verify <target> --address <address>`.",
    });
  }

  return {
    target,
    ...(address === undefined ? {} : { address }),
    ...(chain === undefined ? {} : { chain }),
    ...(verifier === undefined ? {} : { verifier }),
    ...(verifierUrl === undefined ? {} : { verifierUrl }),
    ...(verifierApiKey === undefined ? {} : { verifierApiKey }),
    ...(etherscanApiKey === undefined ? {} : { etherscanApiKey }),
    ...(constructorArgs === undefined ? {} : { constructorArgs }),
    ...(constructorArgsPath === undefined ? {} : { constructorArgsPath }),
    guessConstructorArgs,
    watch,
    showStandardJsonInput,
  };
}

function missingAddress(): string {
  throw new ProjectError({
    code: "verify_address_required",
    message: "No verification address was provided.",
    hint: "Pass `--address <address>` or deploy the target first.",
  });
}
